### PR TITLE
KYLIN-4046 Refine JDBC Source(source.default=8)

### PR DESCRIPTION
**Pull request by hit-lacus**
Currently, the function of ingest data from RDBMS(kylin.source.default=8) to Kylin has some problems , in this patch, I want to fix:

- fix case-sensitive
- fix weak dialect compatibility
- fix mis-use quote character

Manually test passed under 
- MySQL 5.7
- SqlServer 2017

IT test passed

### Configuration

- Sql Server 2017
```
kylin.source.default=8
kylin.source.jdbc.connection-url=jdbc:sqlserver://hdp30-qa:1433;DatabaseName=DemoData
kylin.source.jdbc.driver=com.microsoft.sqlserver.jdbc.SQLServerDriver
kylin.source.jdbc.dialect=mssql
kylin.source.jdbc.user=sa
kylin.source.jdbc.pass=SqlServer_2017
kylin.source.jdbc.filed-delimiter=|
kylin.source.jdbc.sqoop-home=/usr/hdp/current/sqoop-client
```

- MySQL 5.7
```
kylin.source.default=8
kylin.source.jdbc.connection-url=jdbc:mysql://hdp30-qa:13306/DemoData
kylin.source.jdbc.driver=com.mysql.jdbc.Driver
kylin.source.jdbc.dialect=mysql
kylin.source.jdbc.user=root
kylin.source.jdbc.pass=MysqlTest_57
kylin.source.jdbc.filed-delimiter=|
kylin.source.jdbc.sqoop-home=/usr/hdp/current/sqoop-client
```

### Model

```sql

SELECT `USERACTION`.`USER_ID` as `USERACTION_USER_ID` ,
 `USERACTION`.`ACT_TYPE` as `USERACTION_ACT_TYPE` ,
 `USERACTION`.`ITEM_ID` as `USERACTION_ITEM_ID` ,
 `USERACTION`.`LOG_TIME` as `USERACTION_LOG_TIME` ,
 `VIDEOINFO`.`VIDEO_ID` as `VIDEOINFO_VIDEO_ID` ,
 `VIDEOINFO`.`VIDEO_TYPE` as `VIDEOINFO_VIDEO_TYPE` ,
 `VIDEOINFO`.`VIDEO_DURATION` as `VIDEOINFO_VIDEO_DURATION` ,
 `WHERE`.`VIDEO_ID` as `WHERE_VIDEO_ID` ,
 `WHERE`.`SELECT` as `WHERE_SELECT` ,
 `WHERE`.`VIDEO_DURATION` as `WHERE_VIDEO_DURATION` ,
 `USERACTION`.`ACTIVE_MINUTES` as `USERACTION_ACTIVE_MINUTES` ,
 `USERACTION`.`PLAY_TIMES` as `USERACTION_PLAY_TIMES` ,
 `USERACTION`.`PV_ID` as `USERACTION_PV_ID`  
FROM `DBO`.`USERACTION` as `USERACTION` 
  LEFT JOIN `DBO`.`VIDEOINFO` as `VIDEOINFO` ON `USERACTION`.`ITEM_ID` = `VIDEOINFO`.`VIDEO_ID` 
  LEFT JOIN `DBO`.`WHERE` as `WHERE` ON `VIDEOINFO`.`VIDEO_TYPE` = `WHERE`.`SELECT` WHERE 1=1
--



```

### CODE REVIEW

**Comment by asfgit** [comment]
Can one of the admins verify this patch?

**Comment by hit-lacus** [comment]
Fix rat-plugin error.

File: core-metadata/src/main/java/org/apache/kylin/metadata/model/PartitionDesc.java [hunk 1]
```diff
@@ -184,19 +185,26 @@ public TblColRef getPartitionTimeColumnRef() {
...
-        public String buildDateRangeCondition(PartitionDesc partDesc, ISegment seg, SegmentRange segRange) {
+        public String buildDateRangeCondition(PartitionDesc partDesc, ISegment seg, SegmentRange segRange, Function<TblColRef, String> quoteFunc) {
             long startInclusive = (Long) segRange.start.v;
             long endExclusive = (Long) segRange.end.v;
 
             TblColRef partitionDateColumn = partDesc.getPartitionDateColumnRef();
             TblColRef partitionTimeColumn = partDesc.getPartitionTimeColumnRef();
 
+            if (partitionDateColumn != null) {
+                partitionDateColumn.setQuotedFunc(quoteFunc);
```
Comment (nichunen) [inline core-metadata/src/main/java/org/apache/kylin/metadata/model/PartitionDesc.java]: duplicate

File: source-jdbc/src/main/java/org/apache/kylin/source/jdbc/JdbcHiveInputBase.java [hunk 2]
```diff
@@ -229,4 +298,245 @@ protected static String escapeQuotationInSql(String sqlExpr) {
...
+            }
+        }
+        appendJoinStatement(flatDesc, sql, singleLine, metadata, metaMap);
+        appendWhereStatement(flatDesc, sql, singleLine, metadata, metaMap);
+        return sql.toString();
+    }
+
+    private static void appendJoinStatement(IJoinedFlatTableDesc flatDesc, StringBuilder sql, boolean singleLine,
+            IJdbcMetadata metadata, Map<String, String> metaMap) {
+        SourceDialect dialect = metadata.getDialect();
```
Comment (nichunen) [inline source-jdbc/src/main/java/org/apache/kylin/source/jdbc/JdbcHiveInputBase.java]: never used?

File: source-jdbc/src/main/java/org/apache/kylin/source/jdbc/JdbcHiveInputBase.java [hunk 3]
```diff
@@ -229,4 +298,245 @@ protected static String escapeQuotationInSql(String sqlExpr) {
...
+            tblName = quoteIdentifier(tblName, metadata.getDialect());
+        return tblName + "." + colName;
+    }
+
+    /**
+     * @return {SCHEMA_NAME}
+     */
+    static String getSchemaQuoted(Map<String, String> metaMap, String database, IJdbcMetadata metadata,
+            boolean needQuote) {
+        metaMap.get(database.toUpperCase(Locale.ROOT));
```
Comment (nichunen) [inline source-jdbc/src/main/java/org/apache/kylin/source/jdbc/JdbcHiveInputBase.java]: what's this for?

File: source-jdbc/src/main/java/org/apache/kylin/source/jdbc/JdbcHiveInputBase.java [hunk 4]
```diff
@@ -18,38 +18,109 @@
...
+                logger.trace(dumpInfo.toString());
+        }
+
+        /**
+         * Fetch and cache metadata from JDBC API, which will help to resolve
+         * problem of identifier case-sensitive
+         *
+         * @param metadataMap a Map which mapping upper case identifier to real/original identifier
+         */
+        public static void calCachedMeata(Map<String, String> metadataMap, DBConnConf dbconf,
```
Comment (nichunen) [inline source-jdbc/src/main/java/org/apache/kylin/source/jdbc/JdbcHiveInputBase.java]: typo?

File: core-metadata/src/main/java/org/apache/kylin/metadata/model/PartitionDesc.java (hunk 1 above)
Comment (hit-lacus) [reply core-metadata/src/main/java/org/apache/kylin/metadata/model/PartitionDesc.java]: One for partition**Date**Column, another for partition**Time**Column

File: source-jdbc/src/main/java/org/apache/kylin/source/jdbc/JdbcHiveInputBase.java (hunk 2 above)
Comment (hit-lacus) [reply source-jdbc/src/main/java/org/apache/kylin/source/jdbc/JdbcHiveInputBase.java]: Used in Line 326

File: source-jdbc/src/main/java/org/apache/kylin/source/jdbc/JdbcHiveInputBase.java (hunk 4 above)
Comment (hit-lacus) [reply source-jdbc/src/main/java/org/apache/kylin/source/jdbc/JdbcHiveInputBase.java]: I will update.

File: source-jdbc/src/main/java/org/apache/kylin/source/jdbc/JdbcHiveInputBase.java (hunk 2 above)
Comment (hit-lacus) [reply source-jdbc/src/main/java/org/apache/kylin/source/jdbc/JdbcHiveInputBase.java]: I will remove

File: source-jdbc/src/main/java/org/apache/kylin/source/jdbc/JdbcHiveInputBase.java [hunk 5]
```diff
@@ -229,4 +299,232 @@ protected static String escapeQuotationInSql(String sqlExpr) {
...
+        }
+    }
+
+    /**
+     * @return {TABLE_NAME}
+     */
+    static String getTableIdentityQuoted(String database, String table, Map<String, String> metaMap,
+            IJdbcMetadata metadata, boolean needQuote) {
+        String value = fetchValue(database, table, null, metaMap);
+        String[] res = value.split("\\.");
```
Comment (nichunen) [inline source-jdbc/src/main/java/org/apache/kylin/source/jdbc/JdbcHiveInputBase.java]: looks like there is possible that "value" does not contains any dot

Comment (hit-lacus) [reply source-jdbc/src/main/java/org/apache/kylin/source/jdbc/JdbcHiveInputBase.java]: If not contains any dot, res will be a string array with one element, it make no mistake.

**Pull request merged**